File: src/state/GameState.ts
// Central, decoupled game state. Components carry data; systems read/write it.
import type { PositionComponent } from "../components/Position";
import type { HealthComponent } from "../components/Health";
import type { SkillComponent } from "../components/Skills";
import type { InventoryComponent } from "../components/Inventory";
import type { Grid } from "../world/Grid";
import type { FarmPlot } from "../data/Farming";
import type { ActiveClue } from "../data/Clues";
import type { ResourceNode } from "../world/ResourceNode";
import type { BuildingType } from "../data/Buildings";
import type { EquipSlot } from "../data/Items";
import type { AttackStyle } from "../data/Combat";
import { DEFAULT_ATTACK_STYLE } from "../data/Combat";
import { createPosition } from "../components/Position";
import { createHealth } from "../components/Health";
import { createSkillComponent } from "../components/Skills";
import { createInventory } from "../components/Inventory";

/**
 * 10:00 — a new hero opens in clear mid-morning light.
 *
 * dayFactor() ramps 06:30 → 12:30, so 08:00 is only 0.25 daylight (ambient 0.36,
 * sun 0.40) — still visibly gloomy. 10:00 gives 0.58, which reads as daytime
 * without pinning the clock at noon.
 */
export const DAY_START_MINUTE = 10 * 60;

/**
 * Bump when a stored field changes meaning, not merely when fields are added —
 * the sanitizer migrates by version, and a value silently reinterpreted on a new
 * scale is worse than a missing one.
 *
 * 1.1.0 — mastery XP moved off the OSRS skill curve onto its own triangular
 *         curve at 1 XP per unit (was 4).
 */
/**
 * The player character: a young mystic apprentice. Named to sit alongside the
 * settlement's cast — Bram, Wren, Tobias, Eldric — and after the corvid its
 * plum-black robe echoes.
 */
export const DEFAULT_HERO_NAME = "Corvin";

export const SAVE_VERSION = "1.1.0";

export interface TownBuilding {
  id: string;
  type: BuildingType;
  x: number;
  y: number;
  level: number;
}

export interface GameState {
  version: string;
  timestamp: number;
  player: {
    name: string;
    pos: PositionComponent;
    health: HealthComponent;
    skills: SkillComponent;
    inventory: InventoryComponent;
    /** P2 equipment: which item is in each slot (itemId). */
    equipped: Partial<Record<EquipSlot, string>>;
    /** P6: world-map state — discovered points of interest + fast-travel. */
    map: { discovered: string[]; fastTravel: boolean; explored: number[] };
    /** P6.3: quest-log — ids of quests the player has completed (persisted). */
    journal: string[];
    /** The one clue hunt in progress, or null. Persisted; see ClueSystem. */
    clue: ActiveClue | null;
    /** P6.4: meta-progress — kill tallies + unlocked achievements (persisted). */
    meta: { kills: Record<string, number>; achievements: string[]; counters: Record<string, number> };
  };
  world: {
    grid: Grid;
    nodes: Map<string, ResourceNode>;
  };
  /** P-A: in-game clock. Persisted so the time of day survives a reload
   *  instead of resetting to midnight on every launch. */
  clock: { minute: number; day: number };
  town: {
    buildings: TownBuilding[];
    /** P7.3: villager labour — job assignments + the village stock they fill. */
    labour: {
      assignments: Record<string, "woodcutting" | "mining">;
      stock: Record<string, number>;
      acc: Record<string, number>; // ms accrued per villager since last output
      /** P7.6: ms each villager has worked (drives veteran yield tiers). */
      worked: Record<string, number>;
    };
    /** P7.8: market supply/demand — sold/bought volume per item (persisted). */
    market: { supply: Record<string, number>; demand: Record<string, number> };
    /**
     * Farming beds, addressed by index. null = empty.
     *
     * A bed stores only the seed and the epoch ms it was sown, so growth is a
     * function of Date.now() — no accumulated progress to persist, and no offline
     * catch-up pass that could pay out twice.
     */
    farm: { plots: (FarmPlot | null)[] };
  };
  collectionLog: Set<string>;
  /** Player-tunable preferences. Persisted so a choice survives a reload. */
  settings: {
    /**
     * Auto-eat trigger, as a percentage of max HP (0 = never).
     *
     * Was a hardcoded 40%: too eager for a player stretching food across a long
     * trip, too late for one fighting something that can two-shot them.
     */
    autoEatPct: number;
    /** F.1: the fight stance — shifts accuracy/max-hit/defense and which skill trains. */
    attackStyle: AttackStyle;
  };
}

/** Selectable auto-eat thresholds, in percent of max HP. 0 disables it. */
export const AUTO_EAT_STEPS = [0, 20, 30, 40, 50, 60, 75] as const;
export const DEFAULT_AUTO_EAT_PCT = 40;

/** Build a fresh state with a given grid (world wiring added by the world system). */
export function createFreshState(grid: Grid, name: string = DEFAULT_HERO_NAME, startX = 10, startY = 10): GameState {
  return {
    version: SAVE_VERSION,
    timestamp: Date.now(),
    player: {
      name,
      pos: createPosition(startX, startY),
      health: createHealth(100),
      skills: createSkillComponent(),
      inventory: createInventory(),
      equipped: {},
      map: { discovered: [], fastTravel: false, explored: [] },
      journal: [],
      clue: null,
      meta: { kills: {}, achievements: [], counters: {} },
    },
    world: { grid, nodes: new Map() },
    clock: { minute: DAY_START_MINUTE, day: 1 },
    town: {
      buildings: [],
      labour: { assignments: {}, stock: {}, acc: {}, worked: {} },
      market: { supply: {}, demand: {} },
      farm: { plots: [] },
    },
    collectionLog: new Set(),
    settings: { autoEatPct: DEFAULT_AUTO_EAT_PCT, attackStyle: DEFAULT_ATTACK_STYLE },
  };
}

File: unity/Assets/Isoperia/Core/Runtime/State/GameState.cs
using System.Collections.Generic;
using Isoperia.Core.Components;

namespace Isoperia.Core.State
{
    public sealed class TownBuilding
    {
        public string Id;
        public string Type;
        public int X;
        public int Y;
        public int Level = 1;
    }

    /// <summary>
    /// A farming bed stores ONLY the seed and the epoch millisecond it was sown.
    /// Growth is therefore a function of the current time: there is no
    /// accumulated progress to persist, and consequently no offline catch-up pass
    /// that could pay the same growth out twice. Do not "improve" this into a
    /// tick-accumulated system.
    /// </summary>
    public sealed class FarmPlot
    {
        public string SeedId;
        public long PlantedAt;
    }

    public sealed class ActiveClue
    {
        public string Tier;           // "simple" | "hard"
        public double Seed;
        public int Step;
        public List<(int X, int Y)> Sites = new List<(int, int)>();
    }

    public sealed class PlayerState
    {
        public string Name;
        public PositionComponent Pos;
        public HealthComponent Health;
        public SkillComponent Skills;
        public InventoryComponent Inventory;

        /// <summary>Item id per equip slot. Absent key means the slot is empty.</summary>
        public Dictionary<string, string> Equipped = new Dictionary<string, string>();

        public List<string> MapDiscovered = new List<string>();
        public bool MapFastTravel;
        public List<double> MapExplored = new List<double>();

        /// <summary>Ids of completed quests.</summary>
        public List<string> Journal = new List<string>();

        /// <summary>The one clue hunt in progress, or null.</summary>
        public ActiveClue Clue;

        /// <summary>Spent on a combat buff, restored by resting at a Campfire.</summary>
        public int Resolve = GameState.ResolveMax;

        /// <summary>The buff currently drawing from Resolve, or null.</summary>
        public string ActiveBuff;

        /// <summary>Weapon special bar, 0..100. Regenerates anywhere.</summary>
        public int SpecialEnergy = GameState.SpecialMax;

        public Dictionary<string, double> MetaKills = new Dictionary<string, double>();
        public List<string> MetaAchievements = new List<string>();
        public Dictionary<string, double> MetaCounters = new Dictionary<string, double>();
    }

    public sealed class LabourState
    {
        public Dictionary<string, string> Assignments = new Dictionary<string, string>();
        public Dictionary<string, double> Stock = new Dictionary<string, double>();

        /// <summary>Milliseconds accrued per villager since last output.</summary>
        public Dictionary<string, double> Acc = new Dictionary<string, double>();

        /// <summary>Milliseconds each villager has worked; drives veteran yield tiers.</summary>
        public Dictionary<string, double> Worked = new Dictionary<string, double>();
    }

    public sealed class TownState
    {
        public List<TownBuilding> Buildings = new List<TownBuilding>();
        public LabourState Labour = new LabourState();
        public Dictionary<string, double> MarketSupply = new Dictionary<string, double>();
        public Dictionary<string, double> MarketDemand = new Dictionary<string, double>();
        public List<FarmPlot> FarmPlots = new List<FarmPlot>();
    }

    public sealed class SettingsState
    {
        /// <summary>
        /// Auto-eat trigger, as a percentage of max HP; 0 disables. Was a
        /// hardcoded 40%, which is too eager for a player stretching food across
        /// a long trip and too late for one fighting something that two-shots
        /// them.
        /// </summary>
        public int AutoEatPct = GameState.DefaultAutoEatPct;

        /// <summary>Fight stance: shifts accuracy/max-hit/defense and which skill trains.</summary>
        public string AttackStyle = GameState.DefaultAttackStyle;
    }

    /// <summary>Mutable state for one deterministic resource-node id.</summary>
    public sealed class ResourceNodeState
    {
        public int Remaining;
        public long RespawnAt;
    }

    /// <summary>
    /// The central mutable state. Components carry data; systems read and write
    /// it. Port of <c>src/state/GameState.ts</c>.
    ///
    /// Note what is NOT here: the world grid. It is a pure function of the seed
    /// and is regenerated on load rather than stored, which is why saves stay
    /// small. Only mutable world state (occupancy, resource nodes) persists.
    /// </summary>
    public sealed class GameState
    {
        /// <summary>
        /// Bump when a field CHANGES MEANING, not merely when one is added. The
        /// sanitizer migrates by version, and a value silently reinterpreted on a
        /// new scale is worse than a missing one.
        ///
        /// 1.1.0 — mastery XP moved off the OSRS skill curve onto its own
        ///         triangular curve at 1 XP per unit (was 4).
        /// </summary>
        public const string SaveVersion = "1.1.0";

        /// <summary>
        /// A new hero opens in clear mid-morning light. The day curve ramps
        /// 06:30 to 12:30, so 08:00 is only 0.25 daylight and still reads as
        /// gloomy; 10:00 gives 0.58, which is daytime without pinning the clock
        /// at noon.
        /// </summary>
        public const int DayStartMinute = 10 * 60;

        /// <summary>
        /// The player character: a young mystic apprentice. Named to sit
        /// alongside the settlement's cast — Bram, Wren, Tobias, Eldric — and
        /// after the corvid its plum-black robe echoes.
        /// </summary>
        public const string DefaultHeroName = "Corvin";

        public const int ResolveMax = 100;
        public const int SpecialMax = 100;
        public const int DefaultAutoEatPct = 40;
        public const string DefaultAttackStyle = "accurate";

        /// <summary>Selectable auto-eat thresholds, percent of max HP. 0 disables.</summary>
        public static readonly int[] AutoEatSteps = { 0, 20, 30, 40, 50, 60, 75 };

        public string Version = SaveVersion;

        /// <summary>
        /// Epoch milliseconds of the last save. This is what "how long were you
        /// away" is measured from — when it was left unset on load, offline
        /// progression measured from process start, was always ~0, and silently
        /// paid out nothing.
        /// </summary>
        public long Timestamp;

        public PlayerState Player = new PlayerState();
        public TownState Town = new TownState();
        public SettingsState Settings = new SettingsState();

        /// <summary>In-game clock. Persisted so time of day survives a reload.</summary>
        public int ClockMinute = DayStartMinute;
        public int ClockDay = 1;

        /// <summary>
        /// Only changed node state is stored. Node placement and definitions are
        /// regenerated from the Core grid and content tables on every load.
        /// </summary>
        public Dictionary<string, ResourceNodeState> ResourceNodes =
            new Dictionary<string, ResourceNodeState>();

        public HashSet<string> CollectionLog = new HashSet<string>();

        public static GameState CreateFresh(
            string name = DefaultHeroName,
            int startX = 10,
            int startY = 10,
            IItemCatalog catalog = null,
            long nowMs = 0)
        {
            var s = new GameState
            {
                Version = SaveVersion,
                Timestamp = nowMs,
                ClockMinute = DayStartMinute,
                ClockDay = 1,
            };

            s.Player.Name = name;
            s.Player.Pos = PositionComponent.Create(startX, startY);
            s.Player.Health = HealthComponent.Create(100);
            s.Player.Skills = SkillComponent.Create();
            s.Player.Inventory = InventoryComponent.Create(catalog);
            s.Player.Resolve = ResolveMax;
            s.Player.SpecialEnergy = SpecialMax;

            return s;
        }
    }
}
